File: mns-reactor/src/test/java/cz/fb/manaus/reactor/profit/AbstractProfitTest.java
package cz.fb.manaus.reactor.profit;

import cz.fb.manaus.core.model.EventType;
import cz.fb.manaus.core.model.Market;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import cz.fb.manaus.core.test.CoreTestFactory;
import org.apache.commons.lang3.time.DateUtils;
import org.junit.Before;

import java.util.Calendar;
import java.util.Date;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;

import static org.apache.commons.lang3.time.DateUtils.addDays;

public abstract class AbstractProfitTest extends AbstractLocalTestCase {
    protected Date marketDate;
    private Market market;

    @Before
    public void setUp() throws Exception {
        marketDate = addDays(DateUtils.truncate(new Date(), Calendar.MONTH), -10);
        market = CoreTestFactory.newMarket("1", marketDate, CoreTestFactory.MATCH_ODDS);
        market.setEventType(new EventType("1", "soccer"));
    }

    protected void setBetAction(SettledBet... bets) {
        for (int i = 0; i < bets.length; i++) {
            SettledBet bet = bets[i];
            bet.setBetAction(CoreTestFactory.newBetAction(Integer.toString(i + 1), market));
        }
    }

    protected List<SettledBet> generateBets(Optional<Side> requestedSide) {
        List<SettledBet> result = new LinkedList<>();
        for (double price = 1.5d; price < 4; price += 0.02) {
            addSideBets(result, price, Side.LAY, requestedSide);
            addSideBets(result, price + 0.1, Side.BACK, requestedSide);
        }
        setBetAction(result.toArray(new SettledBet[result.size()]));
        return result;
    }

    private void addSideBets(List<SettledBet> result, double price, Side side, Optional<Side> requestedSide) {
        if (requestedSide.orElse(side) == side) {
            result.add(new SettledBet(CoreTestFactory.DRAW, "The Draw", 5d, marketDate, marketDate,
                    new Price(price, 4d, side)));
            result.add(new SettledBet(CoreTestFactory.HOME, "Home", 5d, marketDate, marketDate,
                    new Price(price, 4d, side)));
        }
    }

}

File: mns-reactor/src/test/java/cz/fb/manaus/reactor/profit/AbstractProfitTest.kt
package cz.fb.manaus.reactor.profit

import cz.fb.manaus.core.model.*
import cz.fb.manaus.core.test.AbstractLocalTestCase
import cz.fb.manaus.core.test.CoreTestFactory
import org.apache.commons.lang3.time.DateUtils
import org.apache.commons.lang3.time.DateUtils.addDays
import org.junit.Before
import java.util.*

abstract class AbstractProfitTest : AbstractLocalTestCase() {
    protected lateinit var marketDate: Date
    private lateinit var market: Market

    @Before
    fun setUp() {
        marketDate = addDays(DateUtils.truncate(Date(), Calendar.MONTH), -10)
        market = CoreTestFactory.newMarket("1", marketDate, CoreTestFactory.MATCH_ODDS)
        market.eventType = EventType("1", "soccer")
    }

    protected fun setBetAction(vararg bets: SettledBet) {
        bets.forEachIndexed { i, b ->
            b.betAction = CoreTestFactory.newBetAction(Integer.toString(i + 1), market)
        }
    }

    // TODO not use java Optional in kotlin code
    protected fun generateBets(requestedSide: Optional<Side>): List<SettledBet> {
        val result = mutableListOf<SettledBet>()
        var price = 1.5
        while (price < 4) {
            addSideBets(result, price, Side.LAY, requestedSide)
            addSideBets(result, price + 0.1, Side.BACK, requestedSide)
            price += 0.02
        }
        setBetAction(*result.toTypedArray())
        return result
    }

    private fun addSideBets(result: MutableList<SettledBet>, price: Double, side: Side, requestedSide: Optional<Side>) {
        if (requestedSide.orElse(side) === side) {
            result.add(ModelFactory.newSettled(CoreTestFactory.DRAW, "The Draw", 5.0, marketDate,
                    Price(price, 4.0, side)))
            result.add(ModelFactory.newSettled(CoreTestFactory.HOME, "Home", 5.0, marketDate,
                    Price(price, 4.0, side)))
        }
    }
}
